From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Matthias Maennich <maennich@google.com>
Date: Wed, 21 Oct 2020 11:40:03 +0100
Subject: dwarf-reader: fix lookup for repeated translation unit paths

When using relative compilation unit paths in DWARF, the lookup for an
existing one was done with an incorrect path. In particular, a '/' was
prefixed to the path regardless of whether the compilation_dir is set.
That led to the instantiation of an additional translation unit that and
failed when adding it to the corpus due to violating translation unit
uniqueness. Fix that by correcting the lookup.

	* src/abg-dwarf-reader.cc(build_translation_unit_and_add_to_ir):
	  Fix lookup for potentially already existing translation units.

Reported-by: Dan Albert <danalbert@google.com>
Reviewed-by: Giuliano Procida <gprocida@google.com>
Signed-off-by: Matthias Maennich <maennich@google.com>
---
 src/abg-dwarf-reader.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/abg-dwarf-reader.cc b/src/abg-dwarf-reader.cc
index 8aa5ad1ed8d2..1bbf9740ef52 100644
--- a/src/abg-dwarf-reader.cc
+++ b/src/abg-dwarf-reader.cc
@@ -10928,7 +10928,8 @@ build_translation_unit_and_add_to_ir(read_context&	ctxt,
   // unit.  That is, it's going to be the union of all the translation
   // units of the same path.
   {
-    string abs_path = compilation_dir + "/" + path;
+    const string& abs_path =
+      compilation_dir.empty() ? path : compilation_dir + "/" + path;
     result = ctxt.current_corpus()->find_translation_unit(abs_path);
   }
 
